cumulusci/robotframework/form_handlers.py

```diff
@@ -0,0 +1,162 @@
+
+"""
+FIXME: Something that still needs to be resolved is what to do about
+name clashes. What if a project imports two different libraries and
+each library implements a handler for the same tag and/or a handler
+for which a built-in handler already exists. How do we decide what
+takes precedence. Is this even something we have to worry about?
+
+So, the idea is to find an element, then do something like
+BaseFormHandler.get_handler(element, locator).set("foo")
+
+"""
+
+
+def get_form_handler(element, locator):
+    """Return an instance of handler for the given element
+
+    This will search all subclasses of BaseFormHandler, looking
+    for a class that supports the tag_name of the given element
+    """
+    tag = element.tag_name
+    for subclass in get_all_subclasses(BaseFormHandler):
+        if tag in subclass.tags:
+            # first one wins, but is that the right algorithm? Do
+            # we want to give precedence to external handlers
+            # over our built-in ones?
+            return subclass(element, locator)
+    return None
+
+
+class BaseFormHandler:
+    def __init__(self, element, locator):
+        # concrete element to be handled
+        self.element = element
+
+        # the locator used to get the element, which we use for error
+        # handling.
+        self.locator = locator
+
+    @property
+    def selenium(self):
+        return BuiltIn().get_library_instance("SeleniumLibrary")
+
+    @property
+    def input_element(self):
+        """Returns the actual <input> or <textarea> element inside the element"""
+        elements = self.element.find_elements_by_xpath(
+            ".//*[self::input or self::textarea]"
+        )
+        return elements[0] if elements else None
+
+    # these should be implemented by each of the handlers
+    def set(self, value):
+        pass
+
+    def focus(self):
+        pass
+
+    def clear(self):
+        pass
+
+
+class InputHandler(BaseFormHandler):
+    """An input handler for non-lightning input and textarea form fields
+
+    This is the fallback handler for when we can't find a lightning component.
+    """
+
+    tags = ["input", "textarea"]
+
+    def set(self, value):
+        self.element.send_keys(value)
+
+    def clear(self):
+        """Remove the value in the element"""
+        # oddly, element.clear() doesn't always work.
+        self.selenium.driver.execute_script("arguments[0].value = '';", self.element)
+
+
+class LightningComboboxHandler(BaseFormHandler):
+    """An input handler for comboboxes"""
+
+    tags = ["lightning-combobox"]
+
+    def set(self, value):
+        value_locator = f'//lightning-base-combobox-item[.="{value}"]'
+        wait = 5
+        try:
+            self.input_element.click()
+            self.selenium.wait_until_element_is_visible(value_locator, wait)
+            self.selenium.click_element(value_locator)
+        except Exception as e:
+            logger.debug(f"exception: {e}")
```

Comment:
What's the benefit of logging this? It'll be included in the chained traceback for the exception raised below, no?

Reply:
Good point. I've removed the logging.